integration_tests/tf_distribute_training_test.py

```diff
@@ -50,6 +50,7 @@ def test_model_fit():
             loss=losses.MeanSquaredError(),
             metrics=[metrics.MeanSquaredError()],
         )
+        model.optimizer.build(model.trainable_variables)
```

Comment:
it's because optimizer parameters are created lazily - and it is part of `strategy.run` in model which is in per-replica scope. alternatively, i think this can be added some where in `model.fit` or `model.compile`. i actually have to do a similar thing in another test for metrics as well since there are metrics that also creates variables lazily.

Reply:
But why do the variables need to be created in advance? TensorFlow is supposed to be able to handle lazy variable creation (e.g. at function tracing time). For other backends, we create optimizer and metric variables at the start of `fit()`.

Reply:
i think the probably is the variables created inside function being called in `strategy.run(self.one_step_data)` in trainer. in `strategy.run` it overrides the `strategy.scope()` so variable creation was not properly replicated.

Reply:
We should be fixing that issue first I think -- end users should not be expected to call `build()` themselves. What we could do is start building the optimizer and metrics in `fit()` if a distribution strategy is set. We already do this for the other backends.

Reply:
sounds good i will do that in that in this pr.

Reply:
@fchollet should be good now - let me know if this lgty